### ISSUE TITLE
Using keras.layers.BatchNormalization inside keras.layers.TimeDistributed causes bad validation loss/accuracy

### ISSUE THREAD

**Issue by LukeBolly** [tensorflow/tensorflow#30109]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): tensorflow-gpu 1.13.1
- Python version: 3.6.7
- CUDA/cuDNN version: v10
- GPU model and memory: GTX 980M

**Describe the current behavior**
Using BatchNorm inside a TimeDistributed layer results in poor validation performance. A model trained and validated on identical data can reach 100% train accuracy but never reach that when validating against the same sample. Removing the Batchnorm layer resolves the issue.

See the training results from a toy example:
```
500/500 loss: 1.6910 - sparse_categorical_accuracy: 0.2400 - val_loss: 1.7292 - val_sparse_categorical_accuracy: 0.2000
Epoch 2/100

500/500 loss: 1.2619 - sparse_categorical_accuracy: 0.5600 - val_loss: 1.6112 - val_sparse_categorical_accuracy: 0.2000
Epoch 3/100

500/500 loss: 0.9130 - sparse_categorical_accuracy: 0.9200 - val_loss: 1.5113 - val_sparse_categorical_accuracy: 0.2000
Epoch 4/100

500/500  loss: 0.6420 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.4297 - val_sparse_categorical_accuracy: 0.2000
Epoch 5/100
...
...
...
500/500 loss: 8.2955e-04 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.1042 - val_sparse_categorical_accuracy: 0.2000
Epoch 100/100
```

**Describe the expected behavior**
When training and validating on an identical sample, the model should get the same accuracy during both training and validation. To eliminate the possibility that this is caused by a difference between the running mean/variance and the batchwise mean/variance, a very small value of BatchNorm Momentum was trialed. It was observed that modifying Momentum appears to have no effect. BatchNorm should not affect a model's ability to make good predictions.

**Code to reproduce the issue**
The following code shows the issue. Run it a few times and watch the validation accuracy/loss.
```
from tensorflow import keras
import tensorflow as tf
import numpy as np
import random

time_steps = 5
sample_width = 20
kernel_size = 3
num_filters = 5
num_classes = 5

def build_graph():
    with tf.name_scope("SequenceProcess"):
        sequence_input = keras.layers.Input(shape=(time_steps, sample_width))

        with tf.name_scope("TimeDistributed"):
            sample_input = keras.layers.Input(shape=(sample_width,))

            conv = keras.layers.Reshape([sample_width, 1])(sample_input)
            conv = keras.layers.BatchNormalization(momentum=0.01)(conv)
            conv = keras.layers.Conv1D(num_filters, kernel_size, padding='same')(conv)
            encoded_sample = keras.layers.Reshape((num_filters * sample_width,))(conv)

            sample_model = keras.models.Model(sample_input, encoded_sample)

        processed_samples = keras.layers.TimeDistributed(sample_model)(sequence_input)
        fc = keras.layers.Dense(num_classes, activation='softmax')(processed_samples)

        sequence_model = keras.models.Model(sequence_input, fc)
        sequence_model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['sparse_categorical_accuracy'])

    return sequence_model

def generate_data():
    steps = []
    for i in range(time_steps):
        sample_data = np.random.rand(sample_width)
        steps.append(sample_data)
    steps = np.array(steps)

    labels = []
    for i in range(time_steps):
        label = random.randint(0, num_classes - 1)
        labels.append([label])
    labels = np.array(labels)

    input_data = []
    input_labels = []

    for i in range(1000):
        input_data.append(steps)
        input_labels.append(labels)

    return np.array(input_data), np.array(input_labels)

if __name__ == '__main__':
    data, labels = generate_data()
    model = build_graph()
    model.fit(data, labels, 10, 100, validation_split= 0.5)
```

**Comment by gadagashwini** [comment]
@LukeBolly I am able to reproduce the issue on colab with Tensorflow-gpu 1.13.1 by executing multiple times.

**Comment by LukeBolly** [comment]
Doing some debugging, turns out the running mean and variance aren't being updated:
if you add these lines after the fit
```
    weights = model.layers[1].layer.layers[2].get_weights()
    print('\nmean:\t\t' + str(weights[0][0]) + '\nrunning mean:\t\t' + str(weights[2][0]) + '\n')
    print('variance:\t' + str(weights[1][0]) + '\nrunning variance:\t' + str(weights[3][0]) + '\n')
```

you'll get the output:
```
mean:		1.1386212
running mean:		0.0

variance:	0.06730438
running variance:	1.0
```

https://github.com/keras-team/keras/issues/7466#issuecomment-500002720
https://github.com/keras-team/keras/issues/12400
https://github.com/keras-team/keras/pull/7467

If you've got an idea how I could work around this in the meantime, it would be much appreciated.

**Comment by LukeBolly** [comment]
Updating to 1.14 yielded different but still incorrect results. On the plus side, at least the running mean and variance are being updated now:
```
mean:		1.2015477
running mean:		0.5039695

variance:	-0.19005191
running variance:	0.08481312
```
The current test doesn't catch the error because it just asserts that the values aren't 0 and 1

**Comment by LukeBolly** [comment]
The running mean and variance update in 1.14, though the results still don't seem right
```
Epoch 1/5
500/500 - 2s - loss: 1.2191 - sparse_categorical_accuracy: 0.5640 - val_loss: 0.9704 - val_sparse_categorical_accuracy: 0.6000
Epoch 2/5
500/500 - 0s - loss: 0.2219 - sparse_categorical_accuracy: 1.0000 - val_loss: 0.9911 - val_sparse_categorical_accuracy: 0.4000
Epoch 3/5
500/500 - 0s - loss: 0.0514 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.0707 - val_sparse_categorical_accuracy: 0.4000
Epoch 4/5
500/500 - 0s - loss: 0.0236 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.0686 - val_sparse_categorical_accuracy: 0.4000
Epoch 5/5
500/500 - 0s - loss: 0.0139 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.0246 - val_sparse_categorical_accuracy: 0.4000

gamma:		1.1039916
running mean:		0.18822414

beta:	0.15830743
running variance:	0.63675237
```

Given that the Momentum value is basically 0, shouldn't validation_accuracy be rougly equal to train_accuracy?

**Comment by LukeBolly** [comment]
It turns out momentum is more complicated than I thought. A value of 0.5 makes the train and validation accuracies match.

Using a value of 0 causes the moving averages to overshoot, and you end up with something like this:
```
running mean:		1.0374619		running variance:	-0.82485986
running mean:		0.010333061		running variance:	0.9818245
running mean:		1.0272316		running variance:	-0.80686545
running mean:		0.020461679		running variance:	0.9640094
running mean:		1.017204		running variance:	-0.7892276
running mean:		0.03038919		running variance:	0.94654715
running mean:		1.0073754		running variance:	-0.7719393
running mean:		0.040119976		running variance:	0.9294311
running mean:		0.99774146		running variance:	-0.7549938
running mean:		0.04965794		running variance:	0.9126543
```
which explains the unstable validation loss.

Long story short, don't use a Momentum below 0.5 and don't use BatchNorm inside TimeDistributed in 1.13.
